File: back/src/main/java/xyz/rusin/choretracker/DateHelper.java
package xyz.rusin.choretracker;

import xyz.rusin.choretracker.chores.Recurring;

import java.util.Calendar;
import java.util.Date;
import java.util.List;

public class DateHelper {
    private static final List<Integer> DAYS_OF_WEEK = List.of(
        Calendar.MONDAY,
        Calendar.TUESDAY,
        Calendar.WEDNESDAY,
        Calendar.THURSDAY,
        Calendar.FRIDAY,
        Calendar.SATURDAY,
        Calendar.SUNDAY
    );

    public static Date nextOccurrence(Date from, Recurring recurring) {
        Calendar cal = getNormalizedCalender(from);

        int today = cal.get(Calendar.DAY_OF_WEEK);

        int index = DAYS_OF_WEEK.indexOf(today);
        for (int i = index + 1; i < DAYS_OF_WEEK.size(); i++) {
            int day = DAYS_OF_WEEK.get(i);
            if (isRecurringOn(day, recurring)) {
                cal.add(Calendar.DAY_OF_YEAR, i - index);
                return cal.getTime();
            }
        }
        
        cal.add(Calendar.DAY_OF_YEAR, -(index));
        cal.add(Calendar.WEEK_OF_YEAR, recurring.repeatsEveryWeeks());
        for (int i = 0; i < DAYS_OF_WEEK.size(); i++) {
            int day = DAYS_OF_WEEK.get(i);
            if (isRecurringOn(day, recurring)) {
                cal.add(Calendar.DAY_OF_YEAR, i);
                return cal.getTime();
            }
        }

        return null;
    }

    private static boolean isRecurringOn(int day, Recurring recurring) {
        return switch (day) {
            case Calendar.MONDAY -> recurring.repeatsOnMonday();
            case Calendar.TUESDAY -> recurring.repeatsOnTuesday();
            case Calendar.WEDNESDAY -> recurring.repeatsOnWednesday();
            case Calendar.THURSDAY -> recurring.repeatsOnThursday();
            case Calendar.FRIDAY -> recurring.repeatsOnFriday();
            case Calendar.SATURDAY -> recurring.repeatsOnSaturday();
            case Calendar.SUNDAY -> recurring.repeatsOnSunday();
            default -> false;
        };
    }

    public static Date nextOccurrence(Date from, int days) {
        Calendar cal = getNormalizedCalender(from);
        cal.add(Calendar.DAY_OF_YEAR, days);
        return cal.getTime();
    }

    private static Calendar getNormalizedCalender(Date from) {
        Calendar cal = Calendar.getInstance();
        cal.setTime(from);
        cal.set(Calendar.HOUR_OF_DAY, 0);
        cal.set(Calendar.MINUTE, 0);
        cal.set(Calendar.SECOND, 0);
        cal.set(Calendar.MILLISECOND, 0);
        return cal;
    }

}

File: front/src/browserBackend/dateHelper.ts
import type { DateHelper, Recurring } from "./backendTypes";

export const dateHelper: DateHelper = {
    nextOccurrence(from: Date, recurring: Recurring): Date | null {
        const cal = getNormalizedDate(from);
        const todayDayOfWeek = cal.getDay(); // 0=Sun, 1=Mon...

        // Java DAYS_OF_WEEK: Mon, Tue, Wed, Thu, Fri, Sat, Sun
        // JS equivalents: 1, 2, 3, 4, 5, 6, 0
        const DAYS_OF_WEEK = [1, 2, 3, 4, 5, 6, 0];

        const index = DAYS_OF_WEEK.indexOf(todayDayOfWeek);

        // Check remaining days in current week
        for (let i = index + 1; i < DAYS_OF_WEEK.length; i++) {
            const day = DAYS_OF_WEEK[i];
            if (isRecurringOn(day, recurring)) {
                cal.setDate(cal.getDate() + (i - index));
                return cal;
            }
        }

        // Move to start of next occurrence week
        // Go back to Monday of current week
        cal.setDate(cal.getDate() - index);
        // Add weeks
        cal.setDate(cal.getDate() + (recurring.repeatsEveryWeeks * 7));

        // Check days in that week
        for (let i = 0; i < DAYS_OF_WEEK.length; i++) {
            const day = DAYS_OF_WEEK[i];
            if (isRecurringOn(day, recurring)) {
                cal.setDate(cal.getDate() + i);
                return cal;
            }
        }

        return null;
    },

    nextOccurrenceDays(from: Date, days: number): Date {
        const cal = getNormalizedDate(from);
        cal.setDate(cal.getDate() + days);
        return cal;
    }
};

function getNormalizedDate(from: Date): Date {
    const cal = new Date(from);
    cal.setHours(0, 0, 0, 0);
    return cal;
}

function isRecurringOn(day: number, recurring: Recurring): boolean {
    switch (day) {
        case 1: return recurring.repeatsOnMonday;
        case 2: return recurring.repeatsOnTuesday;
        case 3: return recurring.repeatsOnWednesday;
        case 4: return recurring.repeatsOnThursday;
        case 5: return recurring.repeatsOnFriday;
        case 6: return recurring.repeatsOnSaturday;
        case 0: return recurring.repeatsOnSunday;
        default: return false;
    }
}
